src/client.js

```diff
@@ -0,0 +1,229 @@
+    this.cabalDns = DatDns(cabalDnsOpts)
+  }
+
+  static getDatabaseVersion () {
+    return Cabal.databaseVersion
+  }
+
+  static scrub (key) {
+    return key.replace('cabal://', '').replace('cbl://', '').replace('dat://', '').replace(/\//g, '')
+  }
+
+  resolveName (name, cb) {
+      return this.cabalDns.resolveName(name).then((key) => { 
+          if (!cb) return Client.scrub(key)
+          cb(Client.scrub(key)) 
+      })
+  }
+
+  createCabal () {
+    return this.addCabal(crypto.keyPair().publicKey.toString('hex'))
+  }
+
+  addCabal (key) {
+      let cabalPromise
+      if (typeof key === 'string') {
+          cabalPromise = this.resolveName(key).then((resolvedKey) => {
+            const {temp, dbdir} = this.config
+            const storage = temp ? ram : dbdir + resolvedKey
+            if (!temp) try { mkdirp.sync(path.join(dbdir, resolvedKey, 'views')) } catch (e) {}
+            var db = temp ? memdb() : level(path.join(dbdir, resolvedKey, 'views'))
+            var cabal = Cabal(storage, resolvedKey, {db: db, maxFeeds: this.maxFeeds})
+            this._keyToCabal[resolvedKey] = cabal
+            return cabal
+          })
+      } else {
+          cabalPromise = new Promise((res, rej) => {
+            // a cabal instance was passed in
+            var cabal = key
+            this._keyToCabal[cabal.key] = cabal
+            res(cabal)
+          })
+      }
+      return new Promise((res, rej) => {
+          cabalPromise.then((cabal) => {
+              if (!this.currentCabal) {
+                  this.currentCabal = cabal
+              }
+              cabal.ready(() => {
+                  const details = new CabalDetails(cabal)
+                  this.cabals.set(cabal, details)
+                  cabal.swarm()
+                  this.getCurrentCabal()._emitUpdate()
+                  res(details)
+              })
+          })
+    })
+  }
+
+  focusCabal (key) {
+    const cabal = this._coerceToCabal(key)
+    if (!cabal) {
+      return false
+    }
+    this.currentCabal = cabal
+    return this.cabalToDetails(cabal)
+  }
+
+  removeCabal (key) {
+    const cabal = this._coerceToCabal(key)
+    if (!cabal) {
+      return false
+    }
+
+    const details = this.cabalToDetails(cabal)
+    details._destroy()
+
+    // burn everything we know about the cabal
+    this._keyToCabal[key] = null
+    return this.cabals.delete(cabal)
+  }
+
+  getCabalKeys () {
+    return Object.keys(this._keyToCabal) // ???: sorted?
+  }
+
+  getCurrentCabal () {
+    return this.cabalToDetails(this.currentCabal)
+  }
+
+  getCabalByKey (key) {
+    if (!key) {
+      return this.currentCabal
+    }
+    return this._keyToCabal[key]
+  }
+
+  cabalToDetails (cabal = this.currentCabal) {
+    const details = this.cabals.get(cabal)
+    if (details) {
+      return details
+    }
+    throw new Error('Could not resolve cabal to details, did ya pass in a cabal instance?')
+  }
+
+  connect (cabal = this.currentCabal) {
```

Comment:
Is this used / needed? Adding a cabal (above) automatically starts the swarming process.

Reply:
i think my initial use case was that it could be used for stuff like https://github.com/cabal-club/cabal-headless, where you want to control if you're connected to the swarm or not programmatically